Documentation/cross-platform-performance-and-eventing.md

```diff
@@ -0,0 +1,287 @@
+
+##Introduction
+
+As we bring up CoreCLR on the Linux and Mac platforms, it’s important that we determine how we’ll measure and analyze performance on these platforms.  On Windows we use an event based model that depends on ETW, and we have a good amount of tooling that builds on this approach.  Ideally, we can extend this model to Linux and Mac and re-use much of the Windows tooling.
+
+# Requirements
+
+Ideally, we'd like to have the following functionality on each OS that we bring-up:
+
+- Collection of machine-wide performance data including CPU sampling, threading information (e.g. context switches), and OS specific events / system call tracing.
+- Collection of CLR-specific events that we have today exposed as ETW events.
+- Collection of EventSource events by a default OS-specific collector as we do today with ETW on Windows.
+- User-mode call stacks for both performance and tracing data.
+- Portability of traces across machines, so that analysis can occur off box.
+- Data viewable on collection OS.
+- Stretch: Data can be understood by TraceEvent, which opens up interesting analysis scenarios.
+	- Using PerfView and existing tooling.
+	- Ability to use CAP on non-Windows data.
+
+# Scoping to Reality
+
+While it would be nice to have full parity across Windows, Linux, and Mac with respect to collection and analysis of performance and tracing data, this is likely not going to happen.  Given that we’ve built up a rich set of functionality on Windows, much of which depends on ETW and is specific to the OS, we’re going to see some differences across the other operating systems.
+Our goal should be to do the best job that we can to enable data collection and analysis across the supported operating systems by betting on the right technologies, such that as the landscape across these operating systems evolve, .NET is well positioned to take advantage of the changes without needing to change the fundamental technology choices that we’ve made.  While this choice will likely result in some pain now, it is likely to position us better for the future and align us with the OS communities.
+
+# Linux
+
+## Proposed Design
+
+Given that the performance and tracing tool space on Linux is quite fragmented, there is not one tool that meets all of our requirements.  As such, we'll use two tools when necessary to gather both performance data and tracing data.
+
+**For performance data collection we'll use perf_events**, an in-tree performance tool that provides access to hardware counters, software counters and system call tracing.  Perf_event will be the primary provider of system-wide performance data such as CPU sampling and context switches.
+
+**For tracing we'll use LTTng**.  LTTng supports usermode tracing with no kernelspace requirements.  It allows for strongly typed static events with PID and TID information.  The system is very configurable and allows for enabling and disabling of individual events.
+
+## Tools Considered
```

Comment:
Would be nice to have links to each of the tools so that people have an opportunity to check out the tools for themselves. Same for the references to the tools mentioned earlier.